#!/usr/bin/env Rscript

# Objective : Perform CNV analysis of ONT Adaptive Sampling Panel data
# Created by: xzd
# Created on: 06/19/24


suppressMessages(library(argparser))

Global4Ref="/public/home/xiezhangdong/DB/Reference/hg38/hg38.fasta"
Global4Bed="/public/home/xiezhangdong/projects/FuwaiCHD_Panel/DataFromHospital/Panel_450Genes/CHDPanel_450Genes.hg38.PromoterGeneBody.nochr.10kBin.bed"
Global4BamRef1="/public/home/xiezhangdong/projects/FuwaiCHD_Panel/StandardAnalyse/NA12878/20240708_450Genes_R10_9G/minimap2/NA12878R10/NA12878R10.sort.bam"
Global4BamRef2="/public/home/xiezhangdong/projects/FuwaiCHD_Panel/StandardAnalyse/NA24385/20240613_627Genes_6G/minimap2/NA24385/NA24385.sort.bam"
Value4TransProb="0.0001"

p <- arg_parser("Run Panel CNV anlysis via ExomeDepth")
p <- add_argument(p, "--bamoftest", help="Input BAM file")
p <- add_argument(p, "--bamofref", help="Input BAM file for ref. Sperated by comma if there are multi bams",default=paste(Global4BamRef1,Global4BamRef2,sep=","))
p <- add_argument(p, "--reference", help="File4Reference", default=Global4Ref)
p <- add_argument(p, "--bed", help="bed file", default=Global4Bed)
p <- add_argument(p, "--prefix", help="Directory and file prefix for result")
p <- add_argument(p, "--readcount", help="File for read count if this can be generated by other method like samtools depth. First 4 column names should be chromosome, start, end and GC")
p <- add_argument(p, "--prob", help="Value for transition probability", default=Value4TransProb)


# Import ExomeDepth for data manipulation
suppressMessages(library(ExomeDepth))
myArg <- commandArgs(FALSE)
BinDir = substring(myArg[grep('--file=',myArg)],8)
BinDir = dirname(BinDir)
source(paste0(BinDir,"/class_definition_Revised.R"))

# if there are need for read count
argv <- parse_args(p)
if (is.na(argv$prefix)) { cat("[ Error ] No directory and file prefix specified.\n"); quit(save = "no",status = 2); } else { FilePrefix = argv$prefix }
if (is.na(argv$readcount)) {
	if (is.na(argv$bed)) { cat("[ Error ] No bed specified.\n"); quit(save = "no",status = 2); } else { File4Bed = argv$bed }
	if (is.na(argv$bamoftest)) { cat("[ Error ] No bam of test specified.\n"); quit(save = "no",status = 2); } else { File4BamTest = argv$bamoftest }
	# ref bam should contain at least two samples, otherwise matrix without column names
	if (is.na(argv$bamofref)) { cat("[ Error ] No bam of ref specified.\n"); quit(save = "no",status = 2); } else { File4BamRef = unlist(strsplit(argv$bamofref,",")) }
	if (is.na(argv$reference)) { cat("[ Error ] No reference specified.\n"); quit(save = "no",status = 2); } else { File4Ref = argv$reference }
	List4Bam = c(File4BamTest, File4BamRef)
	cat("[ Info ] Bam of test: ",File4BamTest,"\n")
	cat("[ Info ] Bam of reference: ",paste(File4BamRef,sep = ' '),"\n")
	cat("[ Info ] Bed: ",File4Bed,"\n")
	
	# panel info
	BedInfo = read.table(File4Bed, header = F, sep = '\t', stringsAsFactors = F,comment.char = '')
	
	# read count on panel
	ReadCount = getBamCounts(bed.frame = BedInfo, bam.files = List4Bam, include.chr = FALSE, referenceFasta = File4Ref, min.mapq = 20, read.width = 10000)
	File4ReadCount = paste(FilePrefix, ".ReadCount.txt", sep = '')
	write.table(ReadCount, file = File4ReadCount,col.names = TRUE,row.names = F,sep = "\t",quote = F)
} else {
	File4ReadCount = argv$readcount
	cat("[ Info ] File for read count: ",File4ReadCount,"\n")
	ReadCount = read.table(File4ReadCount, header = T, sep = '\t', stringsAsFactors = F,comment.char = '')
}

# select the appropriate reference(s) or only one
ReadCount4Test = ReadCount[,5]
if (ncol(ReadCount) > 6) {
	ReadCount4Ref = as.matrix(ReadCount[,6:ncol(ReadCount)])
	# n.bins.reduced: the number of exomes to be compared, 10000 is enough
	RefChoiceInfo = select.reference.set(test.counts = ReadCount4Test, reference.counts = ReadCount4Ref, bin.length = (ReadCount$end - ReadCount$start)/1000,n.bins.reduced = 10000)
	RefChoice = RefChoiceInfo$reference.choice
	cat("[ Info ] Selected reference bam are ",paste(RefChoice, sep = ' '),"\n")
} else {
	RefChoice = colnames(ReadCount)[6]
	cat("[ Info ] Selected reference bam is ",RefChoice,"\n")
}
# sum the selected reference samples
# MAR = 1 for row, FUN for sum
RefMatrix = as.matrix(ReadCount[, RefChoice, drop = FALSE])
RefSelectedVector = apply(X = RefMatrix, MAR = 1, FUN = sum)

# map draw
if (TRUE) {
	# test
	tData = ReadCount[,c(1,2,3,5)]
	colnames(tData) = c("Chr","Start","End","Value")
	write.table(tData, file = paste(FilePrefix, ".TestCount.txt", sep = ''), col.names = TRUE, row.names = F, sep = "\t", quote = F)
	
	# relative read count
	tData = ReadCount[,c(1,2,3)]
	colnames(tData) = c("Chr","Start","End")
	tData$Value = ReadCount4Test / (ReadCount4Test + RefSelectedVector)
	tData = tData[!is.na(tData$Value),]
	write.table(tData, file = paste(FilePrefix, ".RelativeCount.txt", sep = ''), col.names = TRUE, row.names = F, sep = "\t", quote = F)
}

## For example: Correlation between reference and tests count is 0.9898, To get meaningful result, this correlation should really be above 0.97.
## for speed improvement: , subset.for.speed = seq(1, nrow(ReadCount), 100)
all.exons <- new('ExomeDepth',test = ReadCount4Test,reference = RefSelectedVector,formula = 'cbind(test, reference) ~ 1')
# call cnv with hidden markov model
cat("[ Transition probability ]",as.numeric(argv$prob),"\n")
all.exons <- CallCNVs(x = all.exons,transition.probability = as.numeric(argv$prob),expected.CNV.length = 50000,chromosome = ReadCount$chromosome,start = ReadCount$start,end = ReadCount$end,name = ReadCount$chromosome)
# cnv result
File4CNV = paste(FilePrefix, ".txt", sep = '')
write.table(all.exons@CNV.calls, file = File4CNV,col.names = TRUE,row.names = F,sep = "\t",quote = F)